### PR TITLE
Link Extension Replace for WikiExport

### PR DESCRIPTION
**Pull request by neildryan**
First pass solution for #100, the argparse section `page.vim:300` wasn't having any effect (arguments seemed to be empty, though that seems like a config issue). Let me know what you think -- happy to make changes.

Some notes:
* This enforces a markdown or mediawiki link style syntax, but I didn't see any indication that other link formats were supported. If they are, I can add other regex or we can go in a different direction.
* This matches links that end with an extension `.md` or `.wiki`, it won't work meaningfully if `g:wiki_link_extension`is empty (this can change as well)

@lervag

### CODE REVIEW

File: autoload/wiki/page.vim [hunk 1]
```diff
@@ -470,9 +472,30 @@ function! s:export(start, end, cfg) abort " {{{1
...
+      let l:link_rx = l:link_rx . g:wiki_link_extension
+      let l:link_rx = l:link_rx . '\(#[^#\(\)\\]\{-}\)\{-})'
+      let l:pattern = '[\1](\2.html\3)'
+    elseif g:wiki_link_target_type ==# 'wiki'
+      let l:link_rx = '\[\[\([^\\\[\]]\{-}\)\.' . g:wiki_link_extension
+      let l:link_rx = l:link_rx . '\(#[^#\\\[\]]\{-}\)'
+      let l:link_rx = l:link_rx . '|\([^\[\]\\]\{-}\)\]\]'
+      let l:pattern = '\[\[\1.html\2|\3\]\]'
+    else
+      echoerr 'Unsupported link type for export replacement'
```
Comment (lervag) [inline autoload/wiki/page.vim:491]: I agree with this approach, but I think we should be explicit about the origin. I.e., mention `g:wiki_link_target_type` in the error message.

File: doc/wiki.txt [hunk 2]
```diff
@@ -662,6 +663,11 @@ the commands are also available as mappings of the form `<plug>(wiki-[name])`.
   -ext EXT~
     Specify extension to use if `fname` is not provided.
 
+  -link_ext_replace~
+    Set to true to replace in-wiki link extensions with html. This enables
+    in-browser wiki navigation. Requires`ext` be set to `html` and for
```
Comment (lervag) [inline doc/wiki.txt:668]: Missing space between `` Requires`ext` ``.

File: doc/wiki.txt [hunk 3]
```diff
@@ -662,6 +663,11 @@ the commands are also available as mappings of the form `<plug>(wiki-[name])`.
...
+    `g:wiki_link_extension` to be non-empty to have any meaningful effect.
```
Comment (lervag) [inline doc/wiki.txt:669]: Change `` `g:wiki_link_extension` `` to `|g:wiki_link_extension|` to enable tag based navigation.

**Comment by lervag** [comment]
Thanks! This looks quite good, I think. I had a couple of minor comments, and I want to test it locally first.

One thought: This feature might develop further, and as such, it could be nice to refactor the code into a separate function. E.g. something like this:

```vim
  " Parse wiki page content
  let l:lines = getline(a:start, a:end)

  if a:cfg.link_ext_replace
    call s:convert_links_html(l:lines)
  endif

  " ...

" And then the function something like this:
function! s:convert_links_html(lines) abort
  " Regex requires that links end with '.', even when wiki_link_extension is
  " empty.
  if g:wiki_link_target_type ==# 'md'
    let l:link_rx = '\[\([^\\\[\]]\{-}\)\](\([^\(\)\\]\{-}\)\.'
    let l:link_rx = l:link_rx . g:wiki_link_extension
    let l:link_rx = l:link_rx . '\(#[^#\(\)\\]\{-}\)\{-})'
    let l:pattern = '[\1](\2.html\3)'
  elseif g:wiki_link_target_type ==# 'wiki'
    let l:link_rx = '\[\[\([^\\\[\]]\{-}\)\.' . g:wiki_link_extension
    let l:link_rx = l:link_rx . '\(#[^#\\\[\]]\{-}\)'
    let l:link_rx = l:link_rx . '|\([^\[\]\\]\{-}\)\]\]'
    let l:pattern = '\[\[\1.html\2|\3\]\]'
  else
    echoerr 'Unsupported link type for export replacement'
    return
  endif
  call map(a:lines, 'substitute(v:val, l:link_rx, l:pattern, ''g'')')
endfunction
```

My example relies on changing the list inplace, as I think that is the most efficient approach, but one could also do it more functionally with `let l:lines = s:convert_links_html(l:lines)`.

Also, I'm curious if it is possible to reuse the regexes defined in `rx.vim`?

**Comment by neildryan** [comment]
Minor changes fixed! Unfortunately, the regex in `rx.vim` don't act with any specificity to the link extension, so I don't think it'd be possible to reuse them without modification.

**Comment by lervag** [comment]
Thanks! I've merged this now. There are room for improvements. For instance, wiki links like `[[name]]` do not work right now. I think it adds a new layer of complexity in that we need to substitutes. I'll leave it to you to decide if it is worth refining and improving this, or if the current version is sufficient.

**Comment by neildryan** [comment]
If someone comes along and has a wiki directory without extensions and wants HTML export, we can change it then. That feels like a pretty small use case; I'll close this for now.

**Pull request merged**